frontend/lib/src/components/widgets/DateInput/DateInput.tsx

```diff
@@ -74,360 +65,320 @@ function datesToStrings(dates: Date[]): string[] {
-          density={DENSITY.high}
-          formatString={dateFormat}
-          mask={isRange ? `${dateMask} – ${dateMask}` : dateMask}
-          placeholder={
-            isRange ? `${element.format} – ${element.format}` : element.format
-          }
-          disabled={disabled}
-          onChange={this.handleChange}
-          onClose={this.handleClose}
-          overrides={{
-            Popover: {
-              props: {
-                placement: PLACEMENT.bottomLeft,
-                overrides: {
-                  Body: {
-                    style: {
-                      border: `${theme.sizes.borderWidth} solid ${colors.borderColor}`,
-                    },
+      setValueWithSource({ value: newValue, fromUi: true })
+      setIsEmpty(!newValue)
+    },
+    [setValueWithSource]
+  )
+
+  const handleClose = useCallback((): void => {
+    if (!isEmpty) return
+
+    const newValue = stringsToDates(element.default)
+    setValueWithSource({ value: newValue, fromUi: true })
+    setIsEmpty(!newValue)
+  }, [isEmpty, element, setValueWithSource])
+
+  return (
+    <div className="stDateInput" data-testid="stDateInput" style={style}>
+      <WidgetLabel
+        label={element.label}
+        disabled={disabled}
+        labelVisibility={labelVisibilityProtoValueToEnum(
+          element.labelVisibility?.value
+        )}
+      >
+        {element.help && (
+          <StyledWidgetLabelHelp>
+            <TooltipIcon
+              content={element.help}
+              placement={Placement.TOP_RIGHT}
+            />
+          </StyledWidgetLabelHelp>
+        )}
+      </WidgetLabel>
+      <UIDatePicker
+        density={DENSITY.high}
+        formatString={dateFormat}
+        mask={element.isRange ? `${dateMask} – ${dateMask}` : dateMask}
+        placeholder={
+          element.isRange
+            ? `${element.format} – ${element.format}`
+            : element.format
+        }
+        disabled={disabled}
+        onChange={handleChange}
+        onClose={handleClose}
+        overrides={{
+          Popover: {
+            props: {
+              placement: PLACEMENT.bottomLeft,
+              overrides: {
+                Body: {
+                  style: {
+                    border: `${theme.sizes.borderWidth} solid ${colors.borderColor}`,
                   },
                 },
               },
             },
-            CalendarContainer: {
-              style: {
-                fontSize: fontSizes.sm,
-                paddingRight: theme.spacing.sm,
-                paddingLeft: theme.spacing.sm,
-                paddingBottom: theme.spacing.sm,
-                paddingTop: theme.spacing.sm,
-              },
+          },
+          CalendarContainer: {
+            style: {
+              fontSize: fontSizes.sm,
+              paddingRight: theme.spacing.sm,
+              paddingLeft: theme.spacing.sm,
+              paddingBottom: theme.spacing.sm,
+              paddingTop: theme.spacing.sm,
```

Comment:
nit: use the destructured property from [here](https://github.com/streamlit/streamlit/pull/9494/files#r1808111702) so that we do not mix the different access ways